oak/server/time/roughtime_client.cc

```diff
@@ -0,0 +1,241 @@
+    freeaddrinfo(address);
+    return Status(asylo::error::GoogleError::INTERNAL, "Failed to create UDP socket.");
+  }
+  if (connect(handle, address->ai_addr, address->ai_addrlen)) {
+    freeaddrinfo(address);
+    close(handle);
+    return Status(asylo::error::GoogleError::INTERNAL, "Failed to open UDP socket.");
+  }
+  freeaddrinfo(address);
+  return handle;
+}
+
+StatusOr<std::string> SendRequest(const RoughtimeServerSpec& server,
+                                  const Nonce<roughtime::kNonceLength>& nonce) {
+  StatusOr<int> create_socket_result;
+  int attempts = 0;
+  do {
+    ++attempts;
+    create_socket_result = CreateSocket(server.host, server.port);
+  } while (!create_socket_result.ok() && attempts <= kServerRetries);
+
+  if (!create_socket_result.ok()) {
+    return Status(asylo::error::GoogleError::INTERNAL,
+                  "Exceeded maximum retries while attempting to connect to " + server.name);
+  }
+
+  auto handle = create_socket_result.ValueOrDie();
+  const std::string request = roughtime::CreateRequest(nonce.data());
+
+  timeval timeout = {};
+  timeout.tv_sec = kTimeoutSeconds;
+  timeout.tv_usec = 0;
+  setsockopt(handle, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
+  ssize_t send_size;
+  do {
+    send_size = send(handle, request.data(), request.size(), 0 /* flags */);
+  } while (send_size == -1 && errno == EINTR);
+  if (send_size < 0 || static_cast<size_t>(send_size) != request.size()) {
+    close(handle);
+    return Status(asylo::error::GoogleError::INTERNAL,
+                  "Network error on sending request to " + server.name);
+  }
+
+  uint8_t receive_buffer[kReceiveBufferSize];
+  ssize_t receive_size;
+  do {
+    receive_size = recv(handle, receive_buffer, kReceiveBufferSize, 0 /* flags */);
+  } while (receive_size == -1 && errno == EINTR);
+  close(handle);
+  if (receive_size == -1) {
+    if (errno == EINTR) {
+      return Status(asylo::error::GoogleError::INTERNAL, "Request timed out for " + server.name);
+    }
+    return Status(asylo::error::GoogleError::INTERNAL, "No response received from " + server.name);
+  }
+
+  return std::string(reinterpret_cast<const char*>(receive_buffer),
+                     static_cast<size_t>(receive_size));
+}
+
+StatusOr<RoughtimeInterval> GetIntervalFromServer(const RoughtimeServerSpec& server) {
+  oak::NonceGenerator<roughtime::kNonceLength> generator;
+  std::string response;
+  Nonce<roughtime::kNonceLength> nonce = generator.NextNonce();
+  ASYLO_ASSIGN_OR_RETURN(response, SendRequest(server, nonce));
+  roughtime::rough_time_t timestamp_microseconds;
+  uint32_t radius_microseconds;
+  std::string error;
+  std::string public_key;
+  if (!absl::Base64Unescape(server.public_key_base64, &public_key)) {
+    return Status(asylo::error::GoogleError::INVALID_ARGUMENT,
+                  absl::StrFormat("Public key for server %s is not a valid base64 encoding.",
+                                  server.name.c_str()));
+  }
```

Comment:
nit: maybe move l.163-l.168 up to the top of the function, so an invalid argument is detected before a request/response happens?